presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/ReservoirSampleState.java

```diff
@@ -0,0 +1,189 @@
+package com.facebook.presto.operator.aggregation.reservoirsample;
+
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.block.BlockBuilder;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.spi.function.AccumulatorState;
+
+import java.util.Arrays;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class ReservoirSampleState implements AccumulatorState {
+
+//    private BlockBuilder blockBuilder;
+    private Block[] samples;
+    private final Type type;
+    private int seenCount;
+
+    private boolean isEmpty = true;
+
+    private final int sampleSize = 200;
+
+    public ReservoirSampleState(Type type)
+    {
+        this.type = requireNonNull(type, "type is null");
+        this.samples = new Block[sampleSize];
+    }
+
+    public ReservoirSampleState(ReservoirSampleState other)
+    {
+        this.type = other.type;
+        this.seenCount = other.seenCount;
+        this.samples = Arrays.copyOf(requireNonNull(other.samples, "samples is null"), other.samples.length);;
+    }
+
+    public void add(Block block, int position)
+    {
+        isEmpty=false;
+        seenCount++;
+        if (seenCount <= sampleSize) {
+            BlockBuilder sampleBlock = type.createBlockBuilder(null, 16);
+            type.appendTo(block, position, sampleBlock);
+            samples[seenCount-1] = sampleBlock.build();
+        }
+        else {
+            int index = ThreadLocalRandom.current().nextInt(0, seenCount);
+            if (index < samples.length) {
+                BlockBuilder sampleBlock = type.createBlockBuilder(null, 16);
+                type.appendTo(block, position, sampleBlock);
+                samples[index] = sampleBlock.build();
+            }
+        }
+    }
+
+    public void addSingleBlock(Block block) {
```

Comment:
```suggestion
    private void addSingleBlock(Block block) {
```

This should be private to avoid usage outside of the class